### PR TITLE
Fix spacing, add ownable, add fallback function for  LuckyNumnber contract

### LINKED ISSUE #373: Fix incomplete code for studio templates rng
See https://github.com/bandprotocol/d3n/pull/370
- Why do we use 4 space here?
- only owner?

### PR DESCRIPTION
**Pull request by prin-r**
Fixes #373

### CODE REVIEW

File: bridges/evm/contracts/example/LuckyNumber.sol [hunk 1]
```diff
@@ -1,56 +1,60 @@
...
-    uint64 public maxNumber;
+contract LuckyNumber is Ownable {
+  bytes32 public codeHash;
+  uint64 public maxNumber;
 
-    uint64 target;
+  uint64 target;
 
-    Bridge bridge = Bridge(0x3e1F8745E4088443350121075828F119075ef641);
+  Bridge bridge = Bridge(0x3e1F8745E4088443350121075828F119075ef641);
```
Comment (sorawit) [inline bridges/evm/contracts/example/LuckyNumber.sol]: Let's make this contract parameter

File: bridges/evm/contracts/example/LuckyNumber.sol [hunk 2]
```diff
@@ -1,56 +1,60 @@
 pragma solidity 0.5.14;
 pragma experimental ABIEncoderV2;
 
+import { Ownable } from "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+
 interface Bridge {
```
Comment (sorawit) [inline bridges/evm/contracts/example/LuckyNumber.sol]: Let's separate this to its own file outside of example.

File: bridges/evm/contracts/example/LuckyNumber.sol [hunk 3]
```diff
@@ -1,56 +1,60 @@
...
-    }
-    function relayAndVerify(bytes calldata data)
-        external
-        returns (VerifyOracleDataResult memory result);
+  struct VerifyOracleDataResult {
+    bytes data;
+    bytes32 codeHash;
+    bytes params;
+  }
+  function relayAndVerify(bytes calldata data)
```
Comment (sorawit) [inline bridges/evm/contracts/example/LuckyNumber.sol]: Make this underscore prefix so we are consistent?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/IBridge.sol [hunk 4]
```diff
@@ -0,0 +1,15 @@
+pragma solidity 0.5.14;
+pragma experimental ABIEncoderV2;
+
+interface IBridge {
```
Comment (sorawit) [inline bridges/evm/contracts/IBridge.sol:4]: 2 space?

File: bridges/evm/contracts/IBridge.sol [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+	/// Helper struct to help the function caller to decode oracle data.
+	struct VerifyOracleDataResult {
+		bytes data;
+		bytes32 codeHash;
+		bytes params;
+	}
+
+	function relayAndVerify(bytes calldata _data)
```
Comment (sorawit) [inline bridges/evm/contracts/IBridge.sol]: Please add docstring here.

File: bridges/evm/contracts/example/DEX.sol [hunk 6]
```diff
@@ -2,128 +2,113 @@ pragma solidity 0.5.14;
 pragma experimental ABIEncoderV2;
 
 import {SafeMath} from "openzeppelin-solidity/contracts/math/SafeMath.sol";
```
Comment (sorawit) [inline bridges/evm/contracts/example/DEX.sol]: Can you add space around SafeMath

File: bridges/evm/contracts/example/DEX.sol [hunk 7]
```diff
@@ -2,128 +2,113 @@ pragma solidity 0.5.14;
...
-
-    bytes32 public codeHash;
-
-    mapping(address => mapping(bytes => uint256)) _balances;
-
-    Bridge bridge = Bridge(0x3e1F8745E4088443350121075828F119075ef641);
-
-    constructor(bytes32 _codeHash) public {
-        codeHash = _codeHash;
+  using SafeMath for uint256;
```
Comment (sorawit) [inline bridges/evm/contracts/example/DEX.sol:9]: Why is the change in this PR? I thought this is about updating LuckyNumber contract?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/example/DEX.sol [hunk 8]
```diff
@@ -1,129 +1,114 @@
...
+    return keccak256(abi.encodePacked(a)) == keccak256(abi.encodePacked(b));
+  }
+
+  function balanceOf(address account, string memory symbol)
+    public
+    view
+    returns (uint256)
+  {
+    bytes memory key;
+    if (strCmp(symbol, "ADA")) {
```
Comment (sorawit) [inline bridges/evm/contracts/example/DEX.sol]: This is crazy inefficient. We are doing `keccak` hash of symbol like 10 times. Why not just store the result in a variable?

File: bridges/evm/contracts/example/DEX.sol [hunk 9]
```diff
@@ -1,129 +1,114 @@
...
-        return keccak256(abi.encodePacked(a)) == keccak256(abi.encodePacked(b));
+    return _balances[account][key];
+  }
+
+  function bytesToPrices(bytes memory _b)
+    public
+    pure
+    returns (uint256, uint256)
+  {
+    require(_b.length >= 16, "INVALID_LENGTH");
```
Comment (sorawit) [inline bridges/evm/contracts/example/DEX.sol]: Why not `==`?

**Review by sorawit — CHANGES_REQUESTED** [review]

**Comment by sorawit** [comment]
@taobun Come help review

File: bridges/evm/contracts/example/DEX.sol [hunk 10]
```diff
@@ -1,129 +1,101 @@
...
-        _balances[msg.sender][result.params] = _balances[msg.sender][result
-            .params]
-            .sub(amount);
-        msg.sender.transfer(ethEarn);
-    }
-
-    function withdraw() public {
-        msg.sender.transfer(address(this).balance);
-    }
+  function withdraw() public {
```
Comment (taobun) [inline bridges/evm/contracts/example/DEX.sol]: onlyOwner

**Review by taobun — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/example/DEX.sol (hunk 10 above)
Comment (prin-r) [reply bridges/evm/contracts/example/DEX.sol]: done

File: bridges/evm/contracts/IBridge.sol [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+
+interface IBridge {
+  /// Helper struct to help the function caller to decode oracle data.
+  struct VerifyOracleDataResult {
+    bytes data;
+    bytes32 codeHash;
+    bytes params;
+  }
+
+  /// @param _data The encoded data for oracle state relay and data verification.
```
Comment (sorawit) [inline bridges/evm/contracts/IBridge.sol:14]: Let's just copy the whole comments from Bridge.sol
```
  /// Performs oracle state relay and oracle data verification in one go. The caller submits
  /// the encoded proof and receives back the decoded data, ready to be validated and used.
  /// @param data The encoded data for oracle state relay and data verification.
```

File: bridges/evm/contracts/example/DEX.sol [hunk 12]
```diff
@@ -1,129 +1,102 @@
...
+    supportedTokens["ETC"] = hex"00000006";
+    supportedTokens["ETH"] = hex"00000007";
+    supportedTokens["LTC"] = hex"00000008";
+    supportedTokens["TRX"] = hex"00000009";
+    supportedTokens["XRP"] = hex"0000000A";
+  }
+
+  function isSupportedToken(string memory a) public view returns(bytes memory, bool) {
+    bytes memory key = supportedTokens[a];
+    return (key, key.length == 4);
```
Comment (sorawit) [inline bridges/evm/contracts/example/DEX.sol]: == 4 is a weird invariant. Let's do != 0 instead?

File: bridges/evm/contracts/example/DEX.sol [hunk 13]
```diff
@@ -1,129 +1,102 @@
...
-            .add(tokenEarn);
-    }
+    _balances[msg.sender][result.params] = _balances[msg.sender][result.params].add(tokenEarn);
+  }
 
-    function sell(uint256 amount, bytes memory _reportPrice) public {
-        Bridge.VerifyOracleDataResult memory result = bridge.relayAndVerify(
-            _reportPrice
-        );
+  function sell(uint256 amount, bytes memory _reportPrice) public {
```
Comment (sorawit) [inline bridges/evm/contracts/example/DEX.sol]: Let's underscore prefix all the parameters.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/IBridge.sol (hunk 11 above)
Comment (prin-r) [reply bridges/evm/contracts/IBridge.sol:14]: done

File: bridges/evm/contracts/example/DEX.sol (hunk 13 above)
Comment (prin-r) [reply bridges/evm/contracts/example/DEX.sol]: done

File: bridges/evm/contracts/example/DEX.sol (hunk 12 above)
Comment (prin-r) [reply bridges/evm/contracts/example/DEX.sol]: done

File: bridges/evm/contracts/IBridge.sol (hunk 11 above)
Comment (prin-r) [reply bridges/evm/contracts/IBridge.sol:14]: done

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**